### PR TITLE
test: fix test failure in case of different default branch

### PR DESCRIPTION
**Pull request by amritghimire**
When the git config has set global `init.defaultBranch` different than master,
a few test cases were failing.
Currently, the test case is assuming the default branch to be set as
master.

This commit will introduce **initial_branch** option to git init call
which will ensure *master* branch is used as default for out test cases.

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Fixes https://github.com/iterative/dvc/issues/6043.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -178,3 +179,90 @@ def custom_template(tmp_dir, dvc):
...
+                "xdg",
+            ),
+        )
+
+    monkeypatch.setenv("GIT_CONFIG_NOSYSTEM", "1")
+
+    with open(os.path.join(home_dir, ".gitconfig"), "wb") as fp:
+        fp.write(
+            b"[user]\n\tname = DVC Tester\n\temail = dvctester@example.com\n"
+            b"[init]\n\tdefaultBranch = master\n"
```
Comment (skshetry) [inline tests/conftest.py:267]: When I do `git init`, it prints the following warning:
```console
$ git init
hint: Using 'master' as the name for the initial branch. This default branch name
hint: is subject to change.
```

So, I don't think we should depend on the hardcoded default branch in our tests **in the future**. To do so, we should introduce `scm.default_branch` property and use that in the tests instead.

File: tests/conftest.py [hunk 2]
```diff
@@ -178,3 +179,90 @@ def custom_template(tmp_dir, dvc):
...
+                "etc",
+                "xdg",
+            ),
+        )
+
+    monkeypatch.setenv("GIT_CONFIG_NOSYSTEM", "1")
+
+    with open(os.path.join(home_dir, ".gitconfig"), "wb") as fp:
+        fp.write(
+            b"[user]\n\tname = DVC Tester\n\temail = dvctester@example.com\n"
```
Comment (skshetry) [inline tests/conftest.py:266]: Do we still need the following config that we set in the GHA?
https://github.com/iterative/dvc/blob/dd0d4f5a2c8dfafb9ae4bf91848b4b0664b75d53/.github/workflows/tests.yaml#L84-L85

**Comment by skshetry** [comment]
@amritghimire ping. Let us know if you are available, we can take over this PR otherwise.

File: tests/conftest.py (hunk 2 above)
Comment (amritghimire) [reply tests/conftest.py:266]: I think this can be removed

**Comment by skshetry** [comment]
I ran into a few issues with the isolation, need to investigate. Most likely this is an issue with our tests how some of the tests don't use `dvc` or `tmp_dir` fixtures at all.

```python
FAILED tests/func/test_ls.py::test_ls_remote_git_only_repo_recursive - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remot...
FAILED tests/func/test_ls.py::test_subrepo[False-git_dir] - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_add.py::test_add_to_remote - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_add.py::test_add_to_remote_invalid_combinations - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ d...
FAILED tests/unit/command/test_add.py::test_add_to_cache_invalid_combinations - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ da...
FAILED tests/unit/command/test_data_sync.py::test_pull - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_data_sync.py::test_push - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_diff.py::test_default - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_diff.py::test_show_hash - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_diff.py::test_show_json - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_diff.py::test_show_json_and_hash - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['...
FAILED tests/unit/command/test_diff.py::test_show_json_hide_missing - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote...
FAILED tests/unit/command/test_diff.py::test_diff_show_markdown_and_hash[None] - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ d...
FAILED tests/unit/command/test_diff.py::test_diff_show_markdown_and_hash[True] - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ d...
FAILED tests/unit/command/test_diff.py::test_diff_show_markdown_and_hash[False] - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ ...
FAILED tests/unit/command/test_diff.py::test_no_changes - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_diff.py::test_hide_missing - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_experiments.py::test_experiments_diff_revs - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['...
FAILED tests/unit/command/test_data_sync.py::test_fetch - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_imp.py::test_import - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_imp.py::test_import_no_exec - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_imp_url.py::test_import_url - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']['minio']
FAILED tests/unit/command/test_imp_url.py::test_failed_import_url - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote']...
FAILED tests/unit/command/test_imp_url.py::test_import_url_no_exec - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remote'...
FAILED tests/unit/command/test_imp_url.py::test_import_url_to_remote - dvc.config.ConfigError: config file error: expected 'url' for dictionary value @ data['remot...
FAILED tests/unit/command/test_imp_url.py::test_import_url_to_remote_invalid_combination - dvc.config.ConfigError: config file error: expected 'url' for dictionary...
```

**Comment by amritghimire** [comment]
@skshetry Can you check if the config file use absolute path for config or something to fetch the global dvc config?

**Comment by skshetry** [comment]
For the record, I have taken over this PR.

**Comment by skshetry** [comment]
For the record, we may just end up isolating git for the first part as done in https://github.com/iterative/scmrepo/blob/9dcd9061cf5e529ebddca5ff9056d26fae1372f6/tests/conftest.py#L37-L62 _stolen_ from this PR.

Isolating `dvc`'s global/system config can be done separately.

**Comment by skshetry** [comment]
@efiop, @dtrifiro this is ready for review, let me know if this solves the issue. 🙂

File: tests/conftest.py [hunk 3]
```diff
@@ -185,3 +186,40 @@ def custom_template(tmp_dir, dvc):
...
+    else:
+        monkeypatch.setenv("HOME", str(home_dir))
+
+    monkeypatch.setenv("GIT_CONFIG_NOSYSTEM", "1")
+    contents = b"""
+[user]
+name=DVC Tester
+email=dvctester@example.com
+[init]
+defaultBranch=master
```
Comment (dtrifiro) [inline tests/conftest.py:219]: Isn't this supposed to be `main`?

**Review by dtrifiro — APPROVED** [review]

File: tests/conftest.py (hunk 3 above)
Comment (skshetry) [reply tests/conftest.py:219]: This is for tests, we always assumed it to be master there. There's https://github.com/iterative/dvc/issues/7005 open to address that.

**Comment by skshetry** [comment]
@dtrifiro, did this fix the issue that you were having?

**Comment by skshetry** [comment]
Merging this, thanks @amritghimire. 

Note that this only provides isolation for git, DVC's own global/system configs are not isolated yet. :)

**Comment by efiop** [comment]
@amritghimire @skshetry thank you guys, outstanding work! 🔥

**Pull request merged**